File: homework2/benjamin_silva/q7_copyTree.py
# Technique: depth-first traversal - pre-order
# Time Complexity: O(n)
# Space Complexity: O(n)
# Time spent: 10 minutes
 
from q3_binarySearchTree import Node
 
 
def copy_tree(root):
    if root is None:
        return None
    
    new_root = Node(root.data)
    new_root.left = copy_tree(root.left)
    new_root.right = copy_tree(root.right)
    return new_root
    
 
 
 
def build_tree(val, left=None, right=None):
    node = Node(val)
    node.left = left
    node.right = right
    return node
 
 
def inorder(node):
    if not node:
        return []
    return inorder(node.left) + [node.data] + inorder(node.right)

original = build_tree(10, build_tree(8, None, build_tree(9)), build_tree(16, build_tree(13), build_tree(17, None, build_tree(20))))
copy = copy_tree(original)
print(inorder(copy))        #expected [8, 9, 10, 13, 16, 17, 20]
print(original is not copy) # expected True 

copy = copy_tree(None)
print(copy)                 # expected None

copy = copy_tree(build_tree(42))
print(inorder(copy))        # expected [42]

copy = copy_tree(build_tree(1, build_tree(2, build_tree(3), None), None))
print(inorder(copy))        #expected [3, 2, 1]

copy = copy_tree(build_tree(1, build_tree(2, build_tree(4), build_tree(5)), build_tree(3, build_tree(6), build_tree(7))))
print(inorder(copy))        # expected [4, 2, 5, 1, 6, 3, 7]

File: homework2/rodrigo_canto/q7_CopyTree.cpp
//Technique - Pre-order Traversal.
//Time Complexity: O(n), where n is the number of nodes in the binary tree.
//Space Complexity: O(n).

#include "bits/stdc++.h"
using namespace std;

struct Node{
    int data;
    Node* left;
    Node* right;
};

Node* createNode(int val){
    
    Node* node = new Node;
    node -> data = val;
    node -> left = nullptr;
    node -> right = nullptr;

    return node;
}

Node* solve_copyTree(Node *curr){

    //When the node does not exist
    if(curr == nullptr) return nullptr;

    Node* new_node = createNode(curr -> data);

    //Left child
    new_node -> left = solve_copyTree(curr -> left);

    //Right child
    new_node -> right = solve_copyTree(curr -> right);

    return new_node;
}

//Function to check if two trees are equal in structure.
//It allows me to see if the solution is working.

bool equal_tree(Node *curr_node, Node* curr_other_node){

    if(curr_node == nullptr && curr_other_node == nullptr) return true;

    if(curr_node == nullptr || curr_other_node == nullptr) return false;

    if(curr_node -> data != curr_other_node -> data) return false;

    return ((equal_tree(curr_node -> left, curr_other_node -> left)) && (equal_tree(curr_node -> right, curr_other_node -> right)));
}

//Function that builds a tree to create inputs.

Node* buildTree(vector<Node*> array){

    for(int i = 0; i < array.size(); ++i){

        Node* curr_node = array[i];

        if(curr_node == nullptr) continue;

        if(2 * i + 1 < array.size()){
            curr_node -> left = array[2 * i + 1];
        }

        if(2 * i + 2 < array.size()){
            curr_node -> right = array[2 * i + 2];
        }
    }

    return array[0];
}

int main(){

    vector<Node*> array = {createNode(10), 
    createNode(8), createNode(16),
    nullptr, createNode(9), createNode(13), createNode(17),
    nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, nullptr, createNode(20)};

    Node* root = buildTree(array);

    Node* new_root = solve_copyTree(root);

    if(equal_tree(root, new_root)) cout << 1 << "\n";
    else cout << 0 << "\n";

    return 0;
}

//Time: 7 minutes